eth/beacon/state_machines/forks/serenity/epoch_transitions.py

```diff
@@ -0,0 +1,224 @@
+        else frozenset.union(*sets_of_current_epoch_boundary_participants)
+    )
+
+    current_epoch_boundary_attesting_balance = Gwei(sum(
+        get_effective_balance(state, index, MAX_DEPOSIT)
+        for index in current_epoch_boundary_attester_indices
+    ))
+    return previous_epoch_boundary_attesting_balance, current_epoch_boundary_attesting_balance
+
+
+def get_total_balance(
+        validator_registry: Sequence[ValidatorRecord],
+        validator_balances: Sequence[Gwei],
+        max_deposit: Ether) -> Gwei:
+
+    active_validator_indices = get_active_validator_indices(validator_registry)
+
+    return total_balance(active_validator_indices, validator_balances, max_deposit)
+
+
+def check_finalization(previous_justified_slot: SlotNumber,
+                       slot: SlotNumber,
+                       justification_bitfield: int,
+                       epoch_length: int)-> bool:
+
+    # Suppose B1, B2, B3, B4 are consecutive blocks and
+    # we are now processing the end of the cycle containing B4.
+
+    # If B4 and is justified using source B3, then B3 is finalized.
+    should_finalize_B3 = (
+        previous_justified_slot == slot - 2 * epoch_length and
+        justification_bitfield % 4 == 3
+    )
+    if should_finalize_B3:
+        return True
+
+    # If B4 is justified using source B2, and B3 has been justified,
+    # then B2 is finalized.
+    should_finalize_B2 = (
+        previous_justified_slot == slot - 3 * epoch_length and
+        justification_bitfield % 8 == 7
+    )
+    if should_finalize_B2:
+        return True
+    # If B3 is justified using source B1, and B1 has been justified,
+    # then B1 is finalized.
+    should_finalize_B1 = (
+        previous_justified_slot == slot - 4 * epoch_length and
+        justification_bitfield % 16 in (15, 14)
+    )
+    if should_finalize_B1:
+        return True
+    return False
+
+
+def process_justification(state: BeaconState, config: BeaconConfig) -> BeaconState:
+    EPOCH_LENGTH = config.EPOCH_LENGTH
+
+    total_balance = get_total_balance(
+        state.validator_registry,
+        state.validator_balances,
+        config.MAX_DEPOSIT,
+    )
+    (
+        previous_epoch_boundary_attesting_balance,
+        current_epoch_boundary_attesting_balance
+    ) = get_epoch_boundary_attesting_balances(state, config)
+
+    previous_justified_slot = state.justified_slot
+    justified_slot = state.justified_slot
+    justification_bitfield = int.from_bytes(state.justification_bitfield, 'big')
+    finalized_slot = state.finalized_slot
+
+    # Add two bits at the right of justification_bitfield. Cap it at 64 bits.
+    justification_bitfield = (justification_bitfield * 2) % TWO_POWER_64
+
+    if 3 * previous_epoch_boundary_attesting_balance >= 2 * total_balance:
+        justification_bitfield |= 2
+        justified_slot = state.slot - 2 * EPOCH_LENGTH
+    if 3 * current_epoch_boundary_attesting_balance >= 2 * total_balance:
```

Comment:
Assuming this isn't intended to be an `elif` then I think it *might* be more appropriate to reverse these two `if` blocks so that we can make them connected using an `elif`.  The goal would be to re-structure this code so that `justification_bitfield` and `justification_slot` (as well as any other variables) are not being overwritten in any locations.

Reply:
Would the desired code look like the follows?
```python
one_epoch_ago_justifiable = 3 * current_epoch_boundary_attesting_balance >= 2 * total_balance
two_epochs_ago_justifiable = 3 * previous_epoch_boundary_attesting_balance >= 2 * total_balance

if one_epoch_ago_justifiable and two_epochs_ago_justifiable:
    justification_bitfield |= 3
    justified_slot = state.slot - EPOCH_LENGTH
elif one_epoch_ago_justifiable and not two_epochs_ago_justifiable :
    justification_bitfield |= 1
    justified_slot = state.slot - EPOCH_LENGTH
elif not one_epoch_ago_justifiable and two_epochs_ago_justifiable:
    justification_bitfield |= 2
    justified_slot = state.slot - 2* EPOCH_LENGTH
```
or maybe
```python
if one_epoch_ago_justifiable:
    justified_slot = state.slot - EPOCH_LENGTH
elif two_epochs_ago_justifiable:
    justified_slot = state.slot - 2* EPOCH_LENGTH

if one_epoch_ago_justifiable and two_epochs_ago_justifiable:
    justification_bitfield |= 3
elif one_epoch_ago_justifiable and not two_epochs_ago_justifiable :
    justification_bitfield |= 1
elif not one_epoch_ago_justifiable and two_epochs_ago_justifiable:
    justification_bitfield |= 2

```

Reply:
Yes, either of these would be :+1: